### PR TITLE
[Core] Update async auth policy lock logic

### PR DESCRIPTION
**Pull request by pvaneck**
Sometimes users might be using trio and its mechanisms for running tasks concurrently. This updates the lock initialization logic to check if the user is in an asyncio event loop and sets the lock accordingly.  If not, assume trio. We don't expect users to be in anything else other than asyncio or trio event loops.

A utility function was added that tries to determine the current async library lock that should be used.

Closes: https://github.com/Azure/azure-sdk-for-python/issues/32968

### CODE REVIEW

File: sdk/core/azure-core/azure/core/pipeline/policies/_authentication_async.py [hunk 1]
```diff
@@ -38,7 +38,20 @@ class AsyncBearerTokenCredentialPolicy(AsyncHTTPPolicy[HTTPRequestType, AsyncHTT
...
+            asyncio.get_running_loop()
+            self._lock = asyncio.Lock()
+        except RuntimeError:
+            # Otherwise, assume we are running in a trio event loop.
+            try:
+                import trio  # pylint:disable=networking-import-outside-azure-core-transport
+
+                self._lock = trio.Lock()
+            except ImportError as err:
+                raise RuntimeError("AsyncBearerTokenCredentialPolicy requires an asyncio or trio event loop.") from err
```
Comment (lmazuel) [inline sdk/core/azure-core/azure/core/pipeline/policies/_authentication_async.py:53]: Customer may be confused. Maybe some thing like:

```suggestion
                raise RuntimeError("This code needs to run as asyncio.run or trio.run.") from err
```

I want to make it actionable by a user receiving the error, what they should do specifically. I'm afraid they won't even know what is the policy name in the first place.

**Review by lmazuel — APPROVED** [review]

**Comment by pvaneck** [comment]
Closing in favor of https://github.com/Azure/azure-sdk-for-python/pull/33307

**Comment by pvaneck** [comment]
Note: Currently trying to resolve CI failures, but they appear to be unrelated to this PR.

**Comment by xiangyan99** [comment]
The change looks good to me. @johanste could you also take a look?

**Comment by pvaneck** [comment]
Note: still need to revisit removing the async lock usage in our policy, and perhaps delegating it to the credential instances.

CI failures due to: https://github.com/Azure/azure-sdk-for-python/issues/33886

**Pull request merged**